plugins/uefi-capsule/fu-uefi-cod-device.c

```diff
@@ -0,0 +1,157 @@
+		return FALSE;
+	if (total_size < 0x3A) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "EFI_CAPSULE_RESULT_VARIABLE_HEADER too small");
+		return FALSE;
+	}
+
+	/* verify guid */
+	if (!fu_memcpy_safe (guid, sizeof(guid), 0x0,		/* dst */
+			     buf, bufsz, 0x08, sizeof(guid), error))
+		return FALSE;
+	guidstr = fwupd_guid_to_string (&guid, FWUPD_GUID_FLAG_MIXED_ENDIAN);
+	if (g_strcmp0 (guidstr, fu_uefi_device_get_guid (device_uefi)) != 0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "wrong GUID, expected %s, got %s",
+			     fu_uefi_device_get_guid (device_uefi),
+			     guidstr);
+		return FALSE;
+	}
+
+	/* get status */
+	if (!fu_common_read_uint32_safe (buf, bufsz, 0x36, &status, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (status == FU_UEFI_DEVICE_STATUS_SUCCESS) {
+		fu_device_set_update_state (device, FWUPD_UPDATE_STATE_SUCCESS);
+		return TRUE;
+	}
+
+	/* something went wrong */
+	if (status == FU_UEFI_DEVICE_STATUS_ERROR_PWR_EVT_AC ||
+	    status == FU_UEFI_DEVICE_STATUS_ERROR_PWR_EVT_BATT) {
+		fu_device_set_update_state (device, FWUPD_UPDATE_STATE_FAILED_TRANSIENT);
+	} else {
+		fu_device_set_update_state (device, FWUPD_UPDATE_STATE_FAILED);
+	}
+	version_str = g_strdup_printf ("%u", fu_uefi_device_get_version_error (device_uefi));
+	tmp = fu_uefi_device_status_to_string (status);
+	if (tmp == NULL) {
+		err_msg = g_strdup_printf ("failed to update to %s",
+					   version_str);
+	} else {
+		err_msg = g_strdup_printf ("failed to update to %s: %s",
+					   version_str, tmp);
+	}
+	fu_device_set_update_error (device, err_msg);
+	return TRUE;
+}
+
+static gboolean
+fu_uefi_cod_device_write_firmware (FuDevice *device,
+				   FuFirmware *firmware,
+				   FwupdInstallFlags flags,
+				   GError **error)
+{
+	FuUefiDevice *self = FU_UEFI_DEVICE (device);
+	guint64 os_indications = EFI_OS_INDICATIONS_FILE_CAPSULE_DELIVERY_SUPPORTED;
+	g_autofree gchar *basename = NULL;
+	g_autofree gchar *cod_path = NULL;
+	g_autofree gchar *esp_path = fu_uefi_device_get_esp_path (self);
+	g_autoptr(GBytes) fw = NULL;
+
+	/* ensure we have the existing state */
+	if ( fu_uefi_device_get_guid (self) == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "cannot update device info with no GUID");
+		return FALSE;
+	}
+
+	/* copy the capsule */
+	fw = fu_firmware_get_bytes (firmware, error);
+	if (fw == NULL)
+		return FALSE;
+	basename = g_strdup_printf ("fwupd-%s.cap", fu_uefi_device_get_guid (self));
+	cod_path = g_build_filename (esp_path, "UpdateCapsule", basename, NULL);
+	if (!fu_common_mkdir_parent (cod_path, error))
+		return FALSE;
+	if (!fu_common_set_contents_bytes (cod_path, fw, error))
+		return FALSE;
+
+	/* assume we can set the runtime variable */
+	if (!fu_efivar_set_data (FU_EFIVAR_GUID_EFI_GLOBAL,
```

Comment:
if the efivarfs is mounted as r/o we need to ignore the setvariable call and just copy the capsule.  
U-Boot has a config option that ignores OsIndications before applying the update.
On the next reboot the Capsule0000 efi variable will be missing, so that's our hint the update didn't run at all.

Reply:
I've added this check.